### ISSUE TITLE
Support expected output in compile-time failure test

### ISSUE THREAD

**Issue by ldionne** [boost-cmake/bcm#1]
For example, I want to be able to check that a specific error message is emitted (a `static_assert`, say). The way I do that right now is by putting the following in the source file:

```c++
// in foo.cpp
int main() {
  // MESSAGE[Incomplete definition of your concept map]
  auto const& map = dyno::complete_concept_map<Fooable, int>(dyno::concept_map<Fooable, int>);
}
```

I then create a `compile_fail` test for `foo.cpp` and that checks whether there's an expected message. See https://github.com/ldionne/dyno/blob/master/cmake/CompileFailTest.cmake for an implementation.

**Comment by pfultz2** [comment]
I think this can be accomplished with the `PASS_REGULAR_EXPRESSION` or `FAIL_REGULAR_EXPRESSION`, but let me create a test for this use case.

**Comment by ldionne** [comment]
I don't think it works (I remember trying this first), and it's convenient to put the error message in the source file, close to the line where the error happens. If you don't put any message, it just checks for failure.

**Comment by pfultz2** [comment]
> I don't think it works 

That looks like what you do in the link you sent.

> it's convenient to put the error message in the source file

Yea, I can add option to parse the output from the source file.

**Comment by ldionne** [comment]
> That looks like what you do in the link you sent.

Sure, but there I don't say `WILL_FAIL`, I just say `PASS_REGULAR_EXPRESSION`. And if I don't want any message, I need to specify `WILL_FAIL` instead. This is backwards, and that's why I think it's nice to have this encapsulated in something like `bcm_test`.

**Comment by pfultz2** [comment]
I actually don't see it as backwards. However, I do see it very useful to add a feature to parse this from the source files.
